File: quint/haskell/lists/p10.hs
-- P10 (*) Run-length encoding of a list.

-- From P09.
pack :: Eq x => [x] -> [[x]]
pack []     = []
pack (x:xs) = (x:y) : pack ys
  where (y, ys) = span (== x) xs

encode :: Eq x => [x] -> [(Int,x)]
-- Map each element on (length, element) in pack x.
encode x = map (\e -> (length e, head e)) (pack x)

File: ben/prolog/P01-P28-lists/p10.pl
% Run-length encoding of a list.

pack_first([], _, [], []). 
pack_first([X|Xs], X, R, [X|Ys]) :-
    pack_first(Xs, X, R, Ys).
pack_first([X|Xs], Y, [X|Xs], []) :- 
    X \= Y.

encode([], []).
encode([Y|Ys], [[Len, Y] |Xs]) :-
    pack_first([Y|Ys], Y, L, X),
    length(X, Len),
    encode(L, Xs).
